src/Microsoft.ML.CpuMath/CpuMathUtils.DotNetCoreApp.cs

```diff
@@ -0,0 +1,362 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics.X86;
+using System;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    public static partial class CpuMathUtils
+    {
+        public const int CbAlign = 16;
+
+        private static bool Compat(AlignedArray a)
+        {
+            Contracts.AssertValue(a);
+            Contracts.Assert(a.Size > 0);
+            return a.CbAlign == CbAlign;
+        }
+
+        private unsafe static float* Ptr(AlignedArray a, float* p)
+        {
+            Contracts.AssertValue(a);
+            float* q = p + a.GetBase((long)p);
+            Contracts.Assert(((long)q & (CbAlign - 1)) == 0);
+            return q;
+        }
+
+        public static void Scale(float a, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count && count <= dst.Length);
+
+            if (Sse.IsSupported)
+            {
+                SseIntrinsics.ScaleU(a, dst, count);
+            }
+            else
+            {
+                for (int i = 0; i < dst.Length; i++)
+                {
+                    dst[i] *= a;
+                }
+            }
+        }
+
+        public static void Scale(float a, float[] dst, int offset, int count)
+        {
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 < count);
+            Contracts.Assert(0 <= offset && offset < dst.Length - count);
+
+            ArraySegment<float> dstSeg = new ArraySegment<float>(dst, offset, count);
+            Scale(a, dstSeg.Array, count);
+        }
+
+        public static void AddScale(float a, float[] src, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(0 < count && count <= src.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(count <= dst.Length);
+
+            if (Sse.IsSupported)
+            {
+                SseIntrinsics.AddScaleU(a, src, dst, count);
+            }
+            else
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    dst[i] += a * src[i];
+                }
+            }
+        }
+
+        public static void AddScale(float a, float[] src, float[] dst, int dstOffset, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(count <= src.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 <= dstOffset && dstOffset < dst.Length);
+            Contracts.Assert(0 < count && count <= dst.Length - dstOffset);
+
+            ArraySegment<float> dstSeg = new ArraySegment<float>(dst, dstOffset, count);
+            AddScale(a, src, dstSeg.Array, count);
```

Comment:
As @eerhardt pointed out above, passing `dstSeg.Array` is not correct, as it just returns the underlying array and doesn't start from `dstOffset`: http://source.dot.net/#System.Private.CoreLib/shared/System/ArraySegment.cs,62

Reply:
Very much agree.  Thanks for pointing it out.  Reference: https://github.com/dotnet/machinelearning/pull/562/files/f0f81a5019a3c8cbd795a970e40d633e9e1770c1#diff-002eee6f7c1e6031c143a2fb39cfe57cR54.